HIP/hip-1046.md

```diff
@@ -2,147 +2,219 @@
+    - A community member can launch a gRPC-Web proxy without external interaction.
+    - A community member can submit a request to have their gRPC-Web proxy added
+      to the network address book.
 
-Any entity wishing to operate a gRPC-Web proxy node would request to have that node added to the network address book via a `nodeCreate` (which, initially, requires council approval), and would manage that node via `nodeUpdate` as needed to maintain current and up-to-date information for the service endpoints of that node.
+## Specification
+
+The dynamic address book stores an entry in state for each node (only consensus,
+currently).  This HIP proposes to extend that “address book” concept to include
+gRPC-Web proxy endpoints as an additional entry in the existing `Node`
+`service_endpoint` list The expectation is that these additional node endpoints
+will be managed with the same `nodeCreate`, `nodeUpdate`, and `nodeDelete`
+transactions that manage the other endpoints for consensus nodes. The only
+changes needed are one additional field in the `ServiceEndpoint` message and
+one enumeration to describe the service endpoint purpose.
+
+Any entity wishing to operate a gRPC-Web proxy would request to have that
+endpoint added to the network address book via a `nodeUpdate` (which, initially,
+requires council approval), and would manage that endpoint via `nodeUpdate` as
+needed to maintain current and up-to-date information for the service endpoint(s)
+of the consensus node behind that proxy.
 
 ### Protobufs
 
 Add a new field to `NodeCreateTransactionBody`.
 
 ```protobuf
-message NodeCreateTransactionBody {
+message ServiceEndpoint {
 ...
     /**
-     * An administrative key controlled by the node operator.
-     */
-    proto.Key admin_key = 7;
-
-    /**
-     * A web proxy for gRPC from non-gRPC clients.
+     * A purpose for this service endpoint.<br/>
+     * This is one of several options, including a gossip endpoint,
+     * a gRPC endpoint, or a gRPC-Web Proxy endpoint, among others.
+     * <p>
+     * This field is REQUIRED.<br/>
+     * This field MUST match the actual purpose of this endpoint.<br/>
+     * There MUST NOT be more than one "gossip" endpoint for each `Node`.<br/>
+     * There MUST NOT be more than one "secure gRPC" endpoint for
+     * each `Node`.<br/>
+     * If a `Node` has a "secure gRPC" endpoint, it SHOULD NOT have an
+     * "insecure gRPC endpoint.
```

Comment:
These statements are not really true. A node can have more than one secure port on different proxy IPs. A node can also have both secure and insecure on the same IP since they are different ports. I imagine in the future with permissionless some nodes might even want multiple secure per IP or multiple insecure per IP. Here's node 0.0.3 service endpoints on mainnet right now:

```json
"service_endpoints":[
{"domain_name":"","ip_address_v4":"34.239.82.6","port":50211},
{"domain_name":"","ip_address_v4":"34.239.82.6","port":50212},
{"domain_name":"","ip_address_v4":"35.237.200.180","port":50211},
{"domain_name":"","ip_address_v4":"35.237.200.180","port":50212}]
```

```

Reply:
It is permitted to do so (SHOULD is a recommendation keyword).  It is not best practice, however.

Reply:
I was commenting on both statements. There is a MUST NOT statement that is not true.

And I disagree with your SHOULD recommendation as it can't be considered a bad practice if every node in production does it. The statements should reflect the reality on mainnet.

Reply:
I see.  That is not ideal (having both secure and insecure endpoints is bad security practice), but I can remove the specification text.